satellite/gracefulexit/endpoint_test.go

```diff
@@ -1207,6 +1207,191 @@ func TestFailureNotFoundPieceHashUnverified(t *testing.T) {
+			MaxThreshold:     4,
+		}
+
+		err := uplinkPeer.UploadWithConfig(ctx, satellite, rs, "testbucket", "test/path", testrand.Bytes(5*memory.KiB))
+		require.NoError(t, err)
+
+		// check that there are no exiting nodes.
+		exitingNodes, err := satellite.DB.OverlayCache().GetExitingNodes(ctx)
+		require.NoError(t, err)
+		require.Len(t, exitingNodes, 0)
+
+		exitingNode, err := findNodeToExit(ctx, planet, 1)
+		require.NoError(t, err)
+
+		// connect to satellite so we initiate the exit.
+		conn, err := exitingNode.Dialer.DialAddressID(ctx, satellite.Addr(), satellite.Identity.ID)
+		require.NoError(t, err)
+		defer ctx.Check(conn.Close)
+
+		client := conn.SatelliteGracefulExitClient()
+
+		c, err := client.Process(ctx)
+		require.NoError(t, err)
+
+		response, err := c.Recv()
+		require.NoError(t, err)
+
+		// should get a NotReady since the metainfo loop would not be finished at this point.
+		switch response.GetMessage().(type) {
+		case *pb.SatelliteMessage_NotReady:
+			// now check that the exiting node is initiated.
+			exitingNodes, err := satellite.DB.OverlayCache().GetExitingNodes(ctx)
+			require.NoError(t, err)
+			require.Len(t, exitingNodes, 1)
+
+			require.Equal(t, exitingNode.ID(), exitingNodes[0].NodeID)
+		default:
+			t.FailNow()
+		}
+		// close the old client
+		require.NoError(t, c.CloseSend())
+
+		// trigger the metainfo loop chore so we can get some pieces to transfer
+		satellite.GracefulExit.Chore.Loop.TriggerWait()
+
+		// make sure all the pieces are in the transfer queue
+		_, err = satellite.DB.GracefulExit().GetIncomplete(ctx, exitingNode.ID(), 1, 0)
+		require.NoError(t, err)
+
+		var notRespondingPiece storj.PieceID
+		var messageCount int
+
+		// We need to label this outer loop so that we're able to exit it from the inner loop.
+		// The outer loop is for sending the request from node to satellite multiple times.
+		// The inner loop is for reading the response.
+	MessageLoop:
+		for messageCount <= maxOrderLimitSendCount {
+			var unknownMsgSent bool
+			c, err := client.Process(ctx)
+			require.NoError(t, err)
+
+			for {
+				response, err := c.Recv()
+				if unknownMsgSent {
+					require.Error(t, err)
+					break
+				} else {
+					require.NoError(t, err)
+				}
+
+				switch m := response.GetMessage().(type) {
+				case *pb.SatelliteMessage_ExitCompleted:
+					break MessageLoop
+				case *pb.SatelliteMessage_TransferPiece:
+					if notRespondingPiece.IsZero() {
+						notRespondingPiece = m.TransferPiece.OriginalPieceId
+					}
+					if m.TransferPiece.OriginalPieceId == notRespondingPiece {
+						messageCount++
+						unknownMsgSent = true
+						// We send an unknown message because we want to fail the
+						// transfer message request we get from the satellite.
+						// This allows us to keep the conn open but repopulate
+						// the pending queue.
+						err = c.Send(&pb.StorageNodeMessage{})
+						require.NoError(t, err)
+						require.NoError(t, c.CloseSend())
+					} else {
```

Comment:
else case not necessary since there are no pieces to successfully transfer in this test